### PR TITLE
[WIP] Countersigning

### PR DESCRIPTION
**Pull request by pdaoust**
For everyone's consideration. Notes/questions to reviewers:

* No images yet. Could probably be useful when describing the session.
* Yes, I do pull out that boogeyman, "double spending", only so I can disabuse devs of the idea that countersigning has anything to do with it. Let me know your thoughts on it.
* Does the story describing the sequence of events involved in a countersigning session actually work? does it flow? is it clear?
* Biggest question: is it actually accurate?
* What's being left out? Does it actually explain what countersigning is for?

### CODE REVIEW

**Comment by Connoropolous** [comment]
hey Paul,

nice, glad this is coming together. I actually haven't done deep dive on countersigning nor used it for anything yet, so I will actually be more learning while reading than anything. 

For everyone, here's a direct link to the published draft of Paul's article: https://deploy-preview-313--fervent-tesla-2a3da2.netlify.app/concepts/10_countersigning

**Comment by Connoropolous** [comment]
@pdaoust It says "The eventual hash of the entry that will be countersigned" and I'm thinking you meant to put that word "eventual" at the end of the sentence?

File: src/concepts/10_countersigning.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+## Key takeaways
+
+* Countersigning allows two or more agents to reach a shared understanding of each other's source chains in order to simultaneously write a new element to all of them.
+* Countersigning is an automated, non-interactive process that happens within a short time window, once all human agreements have been finalized.
+* All participating counterparties must be online for the entire countersigning session.
+* Countersigning begins with the sharing of a preflight, in which one party proposes a session time window, the hash of the entry to be signed, and the list of signers.
+* Countersigning proceeds from preflight to the session once all counterparties have seen each other's preflight signatures as an expression of intent to participate.
+* Before writing the new element to their source chain, each party pre-publishes their header for the element to the DHT, expressing their commitment to write the element.
+* Once the DHT peers have collected the full list of headers, they broadcast them to all parties.
+* Once a party receives all counterparties, headers, they commit the new element to their own source chain and publish it.
```
Comment (jost-s) [inline src/concepts/10_countersigning.md:92]: the comma should be an apostrophe "all counterparties' headers"

File: src/concepts/10_countersigning.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+## Key takeaways
+
+* Countersigning allows two or more agents to reach a shared understanding of each other's source chains in order to simultaneously write a new element to all of them.
+* Countersigning is an automated, non-interactive process that happens within a short time window, once all human agreements have been finalized.
+* All participating counterparties must be online for the entire countersigning session.
+* Countersigning begins with the sharing of a preflight, in which one party proposes a session time window, the hash of the entry to be signed, and the list of signers.
+* Countersigning proceeds from preflight to the session once all counterparties have seen each other's preflight signatures as an expression of intent to participate.
+* Before writing the new element to their source chain, each party pre-publishes their header for the element to the DHT, expressing their commitment to write the element.
+* Once the DHT peers have collected the full list of headers, they broadcast them to all parties.
```
Comment (jost-s) [inline src/concepts/10_countersigning.md:91]: I don't understand this together with the next takeaway.

1. The entry's headers are being published
2. The DHT peers (other than the parties I suppose?) receive the full list of headers and broadcast them (didn't the parties receive them as well?).
3. Once a party receives all headers of all counterparties, they commit the element

File: src/concepts/10_countersigning.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+Here's a more in-depth look at this process. Keep in mind that it's meant to be an automated process, finished within a few seconds while Alice and Bob are both online. Any human agreement (such as negotiating meeting times, contract terms, or transaction details) should happen beforehand.
+
+1. Alice creates a **preflight request** containing:
+    * The parties involved (her and Bob)
+    * The eventual hash of the entry that will be countersigned
+    * The time window in which the countersigning must complete
+    * A stub of the header to go along with the entry, which contains its action type and entry type (currently create-entry and update-entry actions can be countersigned; in the future delete-entry, create-link, and delete-link actions will be too)
+    * Optional application data
+
+2. Alice sends the preflight request to Bob for approval. (You can implement this any way you like, but a [remote call](../8_calls_capabilities/) is probably the most reliable.)
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/concepts/10_countersigning.md:51]: The data structures and the session negotiation stuff are all provided by Holochain; the means of transporting the preflight is the part you have to build yourself. Should I make that clearer somehow?

File: src/concepts/10_countersigning.md (hunk 2 above)
Comment (pdaoust) [reply src/concepts/10_countersigning.md:91]: Yeah, I'm struggling with how to articulate this too. It goes like

1. Parties prepublish their headers to entry hash's neighbourhood
2. Parties wait for neighbourhood to collect everyone's headers and broadcast them to everyone
3. Either:
    * Session times out (parties unlock their chain and throw away the not-yet-committed entry, along with other entries committed in same function call, neighbourhood forgets prepublished headers)
    * Session succeeds (parties write the entries and publish all ops 'for real' this time -- header, entry, and agent activity)

Sounds like better clarity is in order.

File: src/concepts/10_countersigning.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+8. At this point the conductor takes over completely. As with any commit, it attempts to validate the new element. But with a countersigned element, the conductor validates it multiple times --- once from the perspective of each party. This is possible because the countersigning session data structure contains the current states of all signing parties' source chains. Alice can't construct a signature for Bob's header, but she can construct the rest of the header and validate it.
+
+9. Once validation is finished, Alice and Bob both publish the new element, with their own header, to the DHT --- but only to the entry authorities. Because the entry hash is the same for both of them, it'll go to the same set of authorities.
+
+10. Once those authorities have collected all headers, they send the full header set to both Alice and Bob. (If Alice and Bob reach the end of the time window without receiving this header set, however, their conductors roll back and unlock their source chains as if nothing had happened.)
```
[the remark this answers is not in the document]
Comment (pdaoust) [reply src/concepts/10_countersigning.md:67]: Sure, I can write about that. Basically, there is no retry unless you implement it yourself.

**Comment by pdaoust** [comment]
@Connoropolous I can certainly move the 'eventual' thing around :+1:

**Comment by pdaoust** [comment]
@jost-s check out the most recent commit; I've tried to tackle your feedback. Hope I'm not getting too wordy 🤞🏼

File: src/concepts/10_countersigning.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+In order to safely coordinate the writing of a single new element to both of their source chains, Alice and Bob have to be able to:
+
+1. Lock their source chains at a certain point,
+2. Check that the write will be valid from both their perspective and the other party's,
+3. Write the element to their source chains and publish it in a way that lets them back out if something goes wrong,
+4. Wait for the other party to do the same thing, and
+5. Complete the write if everything's successful, or back out if it's not, and unlock their source chains.
+
+Here's a more in-depth look at this process. Keep in mind that it's meant to be an automated process, finished within a few seconds while Alice and Bob are both online. Any human agreement (such as negotiating meeting times, contract terms, or transaction details) should happen beforehand.
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:42]: **Suggest:** ...should happen beforehand, and already be saved to the respective chains as a pre-authorization for this countersigning action.

File: src/concepts/10_countersigning.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+1. Alice creates a **preflight request** containing:
+
+    * The parties involved (her and Bob)
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:46]: **Change:** her --> Alice

The grammar as to whether to use the subjective (she) or objective form (her) is ambiguous here (and therefore an unnecessary distraction) She and Bob are the parties involved. The parties are her and Bob. "Alice" works clearly in both cases.

File: src/concepts/10_countersigning.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    * The hash of the entry that will eventually be countersigned
+    * The time window in which the countersigning must complete
+    * A stub of the header to go along with the entry, which contains its action type and entry type (currently create-entry and update-entry actions can be countersigned; in the future delete-entry, create-link, and delete-link actions will be too)
+    * Optional application data
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:50]: **Suggest**: Optional application data (such as reference to a pre-authorization record showing the human negotiation is complete).

File: src/concepts/10_countersigning.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+2. Alice sends the preflight request to Bob for approval. (You can implement this any way you like, but a [remote call](../8_calls_capabilities/) is probably the most reliable.) At this stage, Bob could use his history with alice, or the optional application data, to determine whether the request is even relevant to him.
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:52]: alice --> Alice

File: src/concepts/10_countersigning.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+8. At this point the conductor takes over completely. As with any commit, it attempts to validate the new element. But with a countersigned element, the conductor validates it multiple times --- once from the perspective of each party. This is possible because the countersigning session data structure contains the current states of all signing parties' source chains. Alice can't construct a signature for Bob's header, but she can construct the rest of the header and validate it.
+
+9. Once validation is finished, Alice and Bob both 'pre-publish' the new element, with their own header, to the DHT --- but only to the entry authorities. This shows their intention to fully commit the element once they see each other do the same. Because the entry hash is the same for both of them, it'll go to the same set of authorities. At this point the data only exists temporarily on the DHT.
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:66]: Suggest: ...go to the same set of DHT authorities.

File: src/concepts/10_countersigning.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+10. Once those authorities have collected all headers, they send the full header set to both Alice and Bob. (If Alice and Bob reach the end of the time window without receiving this header set, however, their conductors discard the new element and unlock their source chains as if nothing had happened. Likewise, the authorities discard the pre-published data. There is no built-in feature to retry an expired session, but it could be built into the application)
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:68]: Suggest: Once any of those authorities have collected all headers, they send the full header set to both Alice and Bob. (However, if Alice and Bob reach the end of the time window without receiving this header set, their conductors discard the new element and unlock their source chains as if nothing had happened. Likewise, the authorities discard the pre-published data if they don't receive the required signatures in the time window. There is no built-in feature to retry an expired session, but it could be built into the application.)

File: src/concepts/10_countersigning.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+## What about double spending?
+
+If you're familiar with blockchains, you might be wondering, _how does countersigning deal with 'double spending'?_ Countersigning, because it allows two or more parties to reach an agreement on something, sounds like it might be an answer to this problem of distributed systems. The plain answer, though, is that **it isn't**.
+
+You could see a blockchain as simply **a mechanism for determining what doesn't exist**, by making a decision about what _does_ exist and rejecting all other possibilities. Many transactional applications, such as currencies and transfer-of title contracts (also known as non-fungible tokens or <abbr>NFTs</abbr>), depend on this feature to prevent one asset from being transferred simultaneously to two different people. The recipient of an asset can be assured that an alternative transaction involving the same asset doesn't secretly exist.
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:76]: transfer-of-title (missing second dash)

File: src/concepts/10_countersigning.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+Holochain, on the other hand, equips all peers to **detect** bad activity, including attempts to 'fork' one's [source chain](../3_source_chain/) and create two simultaneous transactions. Your application can take advantage of [peer witnessing](../4_dht/#a-cloud-of-witnesses) to protect agents from this sort of attack. You can also build extra measures into your application, such as economic disincentives or extra layers of witnessing.
+
+But, as we've mentioned, that's not the job of countersigning. Countersigning merely seals the deal once the parties are ready.
```
Comment (artbrock) [inline src/concepts/10_countersigning.md:80]: Suggest: But since you can already detect double-spend attempts in someone's source chain, that is not the job of countersigning. Countersigning simply allows a group of people to make an identical change to all their chains as a unified operation -- all chains change, or no chains change.

File: src/concepts/11_migration.md [hunk 13]
```diff
@@ -1,4 +1,4 @@
-# 10. Migration: Updating A Network To a New Set of Rules
+# 11. Migration: Updating A Network To a New Set of Rules
 
 !!! note
     This page describes an unimplemented feature. In the meantime, you can implement it yourself by defining your own open and close chain entry types (if needed) and writing migration routines that connect to the old cell to access or import old data.
```
Comment (artbrock) [inline src/concepts/11_migration.md:4]: Suggest: This page includes an unimplemented feature.

**Review by artbrock — CHANGES_REQUESTED** [review]

File: src/concepts/10_countersigning.md (hunk 2 above)
Comment (jost-s) [reply src/concepts/10_countersigning.md:90]: Reading through the page again I find that the detailed steps are clear, and that this part of the takeaways reiterates what has been said before. We could therefore keep takeaways 1, 2 and 3 as well as the last 3 and summarize the ones in-between.

E. g. 
* preflight requests to all parties ensure that a countersigned entry can be written at the given time window
* preflight responses make up the countersigning session data
* all parties validate the entry against the countersigning session data before attempting to commit it
* the entry is first published to the DHT, validated by authorities and then committed to each party's source chain

**Comment by pdaoust** [comment]
Closing in favour of #357 , which incorporates most of the suggestions, adds new content, and also works with Matthew's new 11ty site structure.
